elements/pfelement/src/pfelement.js

```diff
@@ -207,18 +238,200 @@ class PFElement extends HTMLElement {
   }
 
   attributeChangedCallback(attr, oldVal, newVal) {
-    if (!this._pfeClass.cascadingAttributes) {
+    if (attr === "pfe-theme") {
+      this.context_update();
+    }
+
+    if (!this._pfeClass.properties) {
+      return;
+    }
+
+    const propName = this._pfeClass._attr2prop(attr);
+    const propDef = this._pfeClass.properties[propName];
+
+    if (!propDef) {
+      console.warn(`Property ${propName} doesn't exist on ${this._pfeClass.name}`);
       return;
     }
 
-    const cascadeTo = this._pfeClass.cascadingAttributes[attr];
-    if (cascadeTo) {
-      this._copyAttribute(attr, cascadeTo);
+    if (propDef.observer) {
+      this[propDef.observer](this._castPropertyValue(propDef, oldVal), this._castPropertyValue(propDef, newVal));
     }
 
-    if (attr === "pfe-theme") {
-      this.context_update();
+    if (propDef.alias) {
+      console.log(`${attr} set, copying value to aliased attr ${this._pfeClass._prop2attr(propDef.alias)}`);
+      const aliasedAttrVal = this.getAttribute(this._pfeClass._prop2attr(propDef.alias));
+      if (aliasedAttrVal !== newVal) {
+        this[propDef.alias] = newVal;
+      }
+    }
+
+    if (propDef.cascade) {
+      this._copyAttribute(attr, propDef.cascade);
+    }
+  }
+
+  static _validateProperties() {
+    for (let propName in this.properties) {
+      const propDef = this.properties[propName];
+
+      // verify that properties conform to the allowed data types
+      if (!isAllowedType(propDef)) {
+        throw new Error(
+          `Property "${propName}" on ${this.constructor.name} must have type String, Number, or Boolean.`
+        );
+      }
+
+      // verify the property name conforms to our naming rules
+      if (!/^[a-z_]/.test(propName)) {
```

Comment:
Is there a way for our compiler to flag this too so it isn't just a client-side notice but an actual compile-time warning/error?

Reply:
I like it!  This isn't the first time the need for static analysis has come up.  Early on I wanted to be able to `import` or `require` an element in node (in the gulpfile), in order to inject the HTML and CSS strings, rather than doing naive string replacement.  It wasn't possible because HTMLElement is undefined in node, so it threw an error.  I tried a couple of DOM libraries, but ran into issues with both and put it off.

I see two options:

# Stubbed import

Stub out HTMLElement and then import the class.

```js
global.HTMLElement = {};
PfeCard = require("elements/pfe-card/src/pfe-card");
validateProperties(PfeCard);
```

That might work. 

# Full DOM library

```js
require("some-dom-library-that-works");
PfeCard = require("elements/pfe-card/src/pfe-card");
validateProperties(PfeCard);
```